test/e2e/workspacetype/controller_test.go

```diff
@@ -0,0 +1,217 @@
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	clientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	"github.com/kcp-dev/kcp/test/e2e/framework"
+)
+
+func TestClusterWorkspaceTypes(t *testing.T) {
+	t.Parallel()
+
+	type runningServer struct {
+		framework.RunningServer
+		orgKcpClient clientset.Interface
+		orgExpect    framework.RegisterClusterWorkspaceExpectation
+	}
+	var testCases = []struct {
+		name string
+		work func(ctx context.Context, t *testing.T, server runningServer)
+	}{
+		{
+			name: "create a workspace without an explicit type",
+			work: func(ctx context.Context, t *testing.T, server runningServer) {
+				t.Logf("Create a workspace without explicit type")
+				workspace, err := server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Create(ctx, &tenancyv1alpha1.ClusterWorkspace{ObjectMeta: metav1.ObjectMeta{Name: "myapp"}}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create workspace")
+				server.Artifact(t, func() (runtime.Object, error) {
+					return server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Get(ctx, workspace.Name, metav1.GetOptions{})
+				})
+
+				err = server.orgExpect(workspace, ready)
+				require.NoError(t, err)
+
+				t.Logf("Expect workspace to be of Universal type, and no initializers")
+				workspace, err = server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Get(ctx, workspace.Name, metav1.GetOptions{})
+				require.NoError(t, err, "failed to get workspace")
+				require.Equalf(t, workspace.Spec.Type, "Universal", "workspace type is not universal")
+				require.Emptyf(t, workspace.Status.Initializers, "workspace has initializers")
+			},
+		},
+		{
+			name: "create a workspace with an explicit non-existing type",
+			work: func(ctx context.Context, t *testing.T, server runningServer) {
+				t.Logf("Create a workspace with explicit non-existing type")
+				workspace, err := server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaces().Create(ctx, &tenancyv1alpha1.ClusterWorkspace{
+					ObjectMeta: metav1.ObjectMeta{Name: "myapp"},
+					Spec:       tenancyv1alpha1.ClusterWorkspaceSpec{Type: "Foo"},
+				}, metav1.CreateOptions{})
+				require.Error(t, err, "failed to create workspace")
+
+				t.Logf("Create type Foo")
+				_, err = server.orgKcpClient.TenancyV1alpha1().ClusterWorkspaceTypes().Create(ctx, &tenancyv1alpha1.ClusterWorkspaceType{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+					Spec:       tenancyv1alpha1.ClusterWorkspaceTypeSpec{},
+				}, metav1.CreateOptions{})
+				require.NoError(t, err, "failed to create workspace type")
+
+				t.Logf("Create workspace with explicit type Foo again")
+				require.Eventually(t, func() bool {
```

Comment:
Maybe this could be a create call without Eventually? Or is a polling loop suggested for api interaction even when there isn't potential for interaction with controllers? Question applies to other test cases in this file.

Reply:
admission is lister based. It would race.

Reply:
Maybe add a comment documenting this so that future maintainers don't replicate my confusion?

Reply:
done